### PR TITLE
Added a metric to CollectAlignmentSummaryMetrics for the mean length of 3' softclips

### PR DESCRIPTION
**Pull request by yfarjoun**
... considering only those reads that have clipping on that end. This metric might provide an estimate for how much the mean insert-size should change for the number of bases lost to soft-clipping (that which is due to the insert ending) to change significantly.

- I'm not especially happy with the name of the metric. if someone can come up with a  better name I'd be happy to change.

- included tests

### Description

_Give your PR a **concise** yet **descriptive** title_
_Please explain the changes you made here._
_Explain the **motivation** for making this change. What existing problem does the pull request solve?_
_Mention any issues fixed, addressed or otherwise related to this pull request, including issue numbers or hard links for issues in other repos._
_You can delete these instructions once you have written your PR description._

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [ ] Added or modified tests to cover changes and any new functionality
- [ ] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

**Review by meganshand — COMMENTED** [review]
Looks good, just some small comments. Thanks!

File: src/main/java/picard/analysis/AlignmentSummaryMetrics.java [hunk 1]
```diff
@@ -181,4 +185,12 @@
      * PF_ALIGNED_BASES (even though these are not aligned!)
      */
     public double PCT_HARDCLIP;
+
+    /**
+     * The average length of the soft-clipped bases at the 3' end of reads. This could be used as an estimate for
+     * the amount by which the insert-size must be increased in order to obtain a significant reduction in bases
+     * lost due to reading off the end of the insert.
+     */
+    public double AVERAGE_NON_ZERO_3PRIME_SOFTCLIP;
```
[the remark this answers is not in the document]
Comment (yfarjoun) [reply src/main/java/picard/analysis/AlignmentSummaryMetrics.java:194]: :shrug: it seems long and convoluted....was hoping for something pithier (is that really a word!?!)

File: src/test/java/picard/analysis/AlignmentSummaryMetricsCollectorTest.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+                {"2S20M", false, 0},
+                {"2S20M", true, 2},
+                {"20M2S", true, 0},
+                {"20M2S", false, 2},
+
+                {"2S20M5S", false, 5},
+                {"2S20M4S", true, 2},
+                {"4S20M2S", true, 4},
+                {"4S20M2S", false, 2},
```
[the remark this answers is not in the document]
Comment (yfarjoun) [reply src/test/java/picard/analysis/AlignmentSummaryMetricsCollectorTest.java:28]: 👍

File: src/main/java/picard/analysis/AlignmentSummaryMetrics.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (meganshand) [reply src/main/java/picard/analysis/AlignmentSummaryMetrics.java:194]: well then I went in the wrong direction :)

**Comment by yfarjoun** [comment]
I responded to your comments @meganshand

**Review by meganshand — APPROVED** [review]
👍

**Comment by yfarjoun** [comment]
I made another small change...care to look again?

File: src/test/java/picard/analysis/AlignmentSummaryMetricsCollectorTest.java [hunk 3]
```diff
@@ -22,6 +27,9 @@
                 {"20M2S", true, 0},
                 {"20M2S", false, 2},
 
+                {"20M2S4S", false, 6},
```
Comment (meganshand) [inline src/test/java/picard/analysis/AlignmentSummaryMetricsCollectorTest.java:30]: Are these legal cigar strings? Doesn't the S operator have to come at the end? I guess it doesn't matter for this test and I agree with 6 for this length (and 0 for the one below).

**Comment by meganshand** [comment]
Still looks good 👍

File: src/test/java/picard/analysis/AlignmentSummaryMetricsCollectorTest.java (hunk 3 above)
Comment (yfarjoun) [reply src/test/java/picard/analysis/AlignmentSummaryMetricsCollectorTest.java:30]: From the spec:
```
• H can only be present as the first and/or last operation.
• S may only have H operations between them and the ends of the CIGAR string
```

so probably, no. but I wanted to make the code robust to this....

**Pull request merged**